src/gui/guiChatConsole.cpp

```diff
@@ -639,3 +679,109 @@ void GUIChatConsole::setVisible(bool visible)
 	}
 }
 
+// Return how many "ctrl" keycodes successfully found in string, or 0 on fail
+int GUIChatConsole::setupChatClickCtrlKeys(std::string inputline)
+{
+	m_cache_chat_weblink_ctrl_keys.clear();
+
+	irr::EKEY_CODE kc;
+	std::string stemp;
+	size_t startpos = 0, endpos = 0;
+	while(startpos < inputline.size())
+	{
+		// Foreach delimited string,
+		endpos = inputline.find(',', startpos);
+		endpos = std::min(endpos, inputline.find(' ', startpos));
+		// Ignore space/comma
+		if(endpos == startpos)
+		{
+			++endpos;
+		}
+		// Ignore consecutive space/comma
+		else if(endpos - startpos > 1)
+		{
+			// If valid keycode, add it to cached list
+			stemp = inputline.substr(startpos, endpos - startpos);
+			kc = keyname_to_keycode_safemode(stemp.c_str());
+			if(kc != irr::KEY_KEY_CODES_COUNT)
+			{
+				m_cache_chat_weblink_ctrl_keys.push_back(kc);
+			}
+			else
+			{
+				stemp = "Ignoring unknown keycode '" + stemp + "' for chat_weblink_ctrl_keys, check your conf";
+				g_logger.log(LL_WARNING, stemp);
+			}
+		}
+		startpos = endpos;
+	}
+
+	return m_cache_chat_weblink_ctrl_keys.size();
+}
+
+bool GUIChatConsole::isInCtrlKeys(const irr::EKEY_CODE& kc)
+{
+	// To avoid including <algorithm>
+	for(size_t i=0; i<m_cache_chat_weblink_ctrl_keys.size(); ++i)
+	{
+		if(m_cache_chat_weblink_ctrl_keys.at(i) == kc)
+			return true;
+	}
+	return false;
+}
+
+void GUIChatConsole::middleClick(s32 col, s32 row)
+{
+	// Prevent accidental rapid clicking
+	static u32 oldtime = 0;
+	// seriously..
+	u32 newtime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
+
+	// 0.6 seconds should suffice
+	if(newtime - oldtime < 600)
+		return;
+	oldtime = newtime;
+
+	const std::vector<ChatFormattedFragment> & frags = m_chat_backend->getConsoleBuffer().getFormattedLine(row).fragments;
+	std::string weblink = "";         // from frag meta
+
+	// Identify targetted fragment, if exists
+	int ind = frags.size() - 1;
+	while(u32(col - 1) < frags[ind].column)
+	{
+		--ind;
+	}
+	if(ind > -1)
+	{
+		weblink = frags[ind].meta;
+	}
+
+	// Debug help
+	std::string ws;
+	ws = "Middleclick: (" + std::to_string(col) + ',' + std::to_string(row) + ')' + " frags:";
+	for(u32 i=0;i<frags.size();++i)
+	{
+		if(ind == int(i))
+			ws += '*';
+		ws += std::to_string(frags.at(i).column) + '('
+			+ std::to_string(frags.at(i).text.size()) + "),";
+	}
```

Comment:
What does this do? Also `std::stringstream` would be a bit faster and simpler to use, rather than constant re-allocations due to `+=` and manual `to_string` calls.

Also why `frags.at(i)` rather than direct indexing?

Reply:
This is leftover debug output for testing which row/fragment the mouseclick is at. I'll comment it out, but it may be useful someday if the console output code is changed.